apps/passport-server/src/services/devconnectPretixSyncService.ts

```diff
@@ -791,6 +894,225 @@ export class DevconnectPretixSyncService {
+    // This is also how initial OrganizerSyncs are created.
+    for (const id of newOrgIds) {
+      const config = devconnectPretixConfig.organizers.find(
+        (org) => org.id === id
+      ) as DevconnectPretixOrganizerConfig;
+      const org = new OrganizerSync(
+        this.dbQueue, // To allow sync jobs to avoid concurrent DB access
+        config,
+        DevconnectPretixSyncService.PRETIX_RATE_PER_MINUTE,
+        this.pretixAPI,
+        this.rollbarService,
+        this.db
+      );
+      this.organizers.set(id, org);
+    }
+
+    for (const id of removedOrgIds) {
+      // @todo cleanup any organizers which have pending requests?
+      this.organizers.delete(id);
+    }
+  }
+
+  /**
+   * Download Pretix state, and apply a diff to our state so that it
+   * reflects the state in Pretix.
+   */
+  private async sync(): Promise<void> {
+    return traced(NAME, "sync", async (span) => {
+      span?.setAttribute("organizers_count", this.organizers.size);
+
+      const syncStart = Date.now();
+      const organizerPromises = [];
+      this.syncResults.clear();
+
+      // Attempt to run each organizer job in parallel
+      // Internally the organizers will use a queue to avoid excessive
+      // concurrent requests to the DB.
+      for (const [id, organizer] of this.organizers.entries()) {
+        organizerPromises.push(
+          (async (): Promise<SyncResult> => {
+            try {
+              return await organizer.run();
+            } catch (e) {
+              logger(
+                `[DEVCONNECT PRETIX] Error encounted when synchronizing organizer ${id}`,
+                e
+              );
+              setError(e, span);
+              this.rollbarService?.reportError(e);
+
+              // If we're catching an exception here, it really *should*
+              // have a .cause of type SyncErrorCause, but we can't type-
+              // check this.
+              if (
+                e instanceof Error &&
+                e.cause &&
+                e.cause instanceof Object &&
+                Object.hasOwn(e.cause, "success") &&
+                Object.hasOwn(e.cause, "phase") &&
+                Object.hasOwn(e.cause, "error") &&
+                Object.hasOwn(e.cause, "organizerId")
+              ) {
+                return e.cause as SyncErrorCause;
+              } else {
+                // This should never happen, but let's handle it anyway
+                logger(
+                  `[DEVCONNECT PRETIX] Unknown error encounted when synchronizing organizer ${id}`,
+                  e
+                );
+                return {
+                  success: false,
+                  error: new Error(
+                    `Unknown error encounted when synchronizing organizer ${id}`
+                  ),
+                  phase: "saving",
+                  organizerId: id
+                };
+              }
+            }
+          })()
```

Comment:
this self-invoking function is difficult for me to follow

Reply:
Good point. It should be a named function.